webrtc.html

```diff
@@ -1873,28 +1800,100 @@ <h4>Set the RTCSessionDescription</h4>
+                              </li>
+                              <li>
+                                <p>Remove <var>transceiver</var> from <var>connection</var>'s
+                                <a href="#transceivers-set">set of transceivers</a>.</p>
+                              </li>
+                              <li>
+                                <p><a>Set the associated remote streams</a> given
+                                <var>transceiver</var>.<a>[[\Receiver]]</a>, <var>msids</var>
+                                as the empty list, <var>addList</var>, and <var>removeList</var>.</p>
+                              </li>
+                              <li>
+                                <p>If the <var>transceiver</var>'s <a>[[\FiredDirection]]</a>
+                                is either <code>"sendrecv"</code> or <code>"recvonly"</code>,
+                                <a>process the removal of a remote track</a> given
+                                <var>transceiver</var> and <var>muteTracks</var>.</p>
+                              </li>
+                            </ol>
+                          </li>
+                          <li>
+                            <p>Otherwise, (if the <var>transceiver</var> was not just removed)
+                            run the following steps:</p>
+                            <ol>
+                              <li>
+                                <p>If the <var>transceiver</var>'s <a>[[\FiredDirection]]</a>
+                                is either <code>"sendonly"</code> or <code>"inactive"</code>
+                                and <var>transceiver</var>'s <a>[[\CurrentDirection]]</a> is
+                                either <code>"sendrecv"</code> or <code>"recvonly"</code>,
+                                <a>process the addition of a remote track</a> given
+                                <var>transceiver</var> and <var>trackEventInits</var> and
+                                set the <var>transceiver</var>'s <a>[[\Receptive]]</a> slot
+                                to <code>true</code>.</p>
+                              </li>
+                              <li>
+                                <p>If the <var>transceiver</var>'s <a>[[\FiredDirection]]</a>
+                                is either <code>"sendrecv"</code> or <code>"recvonly"</code>
+                                and <var>transceiver</var>'s <a>[[\CurrentDirection]]</a> is
+                                either <code>"sendonly"</code>, <code>"inactive"</code> or
+                                <code>null</code>, <a>process the removal of a remote track</a>
+                                given <var>transceiver</var> and <var>muteTracks</var> and
+                                set the <var>transceiver</var>'s <a>[[\Receptive]]</a> slot
+                                to <code>false</code>.</p>
+                              </li>
```

Comment:
For both of the if-else parts above we're forgetting to update [[FiredDirection]].

We're also forgetting the [[Receptive]] slot...

these are the core parts that I sought to reuse; there's a lot of overlap. Getting too hard to keep track...

So I propose we break out a higher-level algorithm here under [§ 5.1.1. Processing Remote MediaStreamTracks](https://w3c.github.io/webrtc-pc/#processing-remote-mediastreamtracks), that takes care of calling all the sub algorithms in that section, as well as updating [[FiredDirection]] and [[Receptive]].

E.g. called simply: **process remote tracks** with _transceiver_, _newDirection_, _msids_, _addList_, _removeList_ and _trackEventInits_.

It then takes care of (loosely copied):
 1. Let _msids_ be a list of the MSIDs that the media description indicates transceiver.[[Receiver]].[[ReceiverTrack]] is to be associated with, or an empty list if no association is indicated.
 2. [Set the associated remote streams](https://w3c.github.io/webrtc-pc/#set-associated-remote-streams) given transceiver.[[Receiver]], msids, addList, and removeList.
 3. If direction is "sendrecv" or "recvonly" and transceiver.[[FiredDirection]] is neither "sendrecv" nor "recvonly", or the previous step increased the length of addList, [process the addition of a remote track](http://w3c.github.io/webrtc-pc/#process-remote-track-addition) for the media description, given transceiver and trackEventInits.
 4. If direction is "sendonly" or "inactive", set transceiver.[[Receptive]] to false.
 5. If direction is "sendonly" or "inactive", and transceiver.[[FiredDirection]] is either "sendrecv" or "recvonly", [process the removal of a remote track](http://w3c.github.io/webrtc-pc/#process-remote-track-removal) with transceiver and muteTracks.
 6. Set transceiver.[[FiredDirection]] to direction.

Hope that helps!

Reply:
Also, minor nit, I'd like to see us be consistent and use _"with"_ at the call-site and _"given"_ at the head of the algorithm. I think that's right, but let me know if I got that backwards.

Reply:
> For both of the if-else parts above we're forgetting to update [[FiredDirection]].

I updated the if-removing-transceiver part to set [[FiredDirection]] to "inactive", though this slot is unused after removal, so it shouldn't matter.
In the otherwise part I added to set it to [[CurrentDirection]].

> We're also forgetting the [[Receptive]] slot...

This is not forgotten; the if-removing-transceiver part takes care of [[Receptive]] inside stopping the transceiver and in the otherwise part this is explicitly set to true or false depending on processing addition or removal of the track.

> E.g. called simply: **process remote tracks** with _transceiver_, _newDirection_, _msids_, _addList_, _removeList_ and _trackEventInits_.

Hang on, this moves everything around, let me revisit this suggestion in a separate commit. Or perhaps a separate PR even?

Reply:
There are still some bugs here. Filing a PR fix now.